src/rslice.rs

```diff
@@ -0,0 +1,184 @@
+use libc::{c_char, size_t};
+use std::convert::{TryFrom, TryInto};
+use std::marker::PhantomData;
+
+/// A read-only view on a Rust byte slice.
+///
+/// This is used to pass data from crustls to callback functions provided
+/// by the user of the API.
+/// `len` indicates the number of bytes than can be safely read.
+///
+/// The memory exposed is available as specified by the function
+/// using this in its signature. For instance, when this is a parameter to a
+/// callback, the lifetime will usually be the duration of the callback.
+/// Functions that receive one of these must not dereference the data pointer
+/// beyond the allowed lifetime.
+#[repr(C)]
+pub struct rustls_slice_bytes<'a> {
+    pub data: *const u8,
+    pub len: size_t,
+    phantom: PhantomData<&'a [u8]>,
+}
+
+impl<'a> From<&'a [u8]> for rustls_slice_bytes<'a> {
+    fn from(s: &[u8]) -> Self {
+        rustls_slice_bytes {
+            data: s.as_ptr() as *const u8,
+            len: s.len() as size_t,
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// An owned Vec<rustls_slice_bytes>, where the inner slices have lifetime 'a.
+/// If we want to share a view of a `Vec<Vec<_>>` with C, we can't do it
+/// directly because `Vec` is not #[repr(C)]. So we build a new Vec containing
+/// rustls_slice_bytes<'a> as an owned object. We can then expose slices of
+/// that Vec to C.
+pub(crate) struct VecSliceBytes<'a>(Vec<rustls_slice_bytes<'a>>);
+
+impl<'a> VecSliceBytes<'a> {
```

Comment:
I'm a bit confused on how this is intended to be used, can you sketch out an example?

I also don't love needing to allocate for this. It seems preferable to keep the original `Vec<Vec<u8>>` and have C functions to extract a `rustls_slice_bytes` at a given index?

Reply:
One example here: https://github.com/abetterinternet/crustls/pull/51/files#diff-7f93c4e263c4e9ec748f804c7fd04a3b2fde86ffd741fb5516d67e1097bae4c1R149-R183

And one here: https://github.com/abetterinternet/crustls/pull/50/files#diff-48d27698cc708c7efe770fd897e4885efaa0a15cae30d54936068603ba03bbd9R696-R711

Good point on accessor functions! I recall that you had recommended that before and I forgot to pursue it. I'll give that a try.

Reply:
For the accessor version, I'm thinking of a function signature like this:

```
pub extern "C" fn rustls_slice_slice_bytes_get(
    input: rustls_slice_slice_bytes, n: usize) -> rustls_slice_bytes {
```

But this requires there be some sort of sentinel value of `rustls_slice_bytes` that means "n was too big." I guess that can be `rustls_slice_bytes{data: null(), len: 0}`. Interested in other ideas! I could also do:

```
pub extern "C" fn rustls_slice_slice_bytes_get(
    input: rustls_slice_slice_bytes, n: usize, out: *mut rustls_slice_bytes) -> rustls_result {
```

But that feels pretty awkward to use in a loop.

Reply:
I agree on the accessor recommendation. They would allow encapsulating some memory layouts, cutting down on the number of types needed on the Rust size, plus calls to the accessors could be inlined when using a cross-language LTO-capable toolchain, making this a cheap abstraction.

The first accessor signature looks good to me, NULL checks seem like idiomatic C IMHO.